cmd/clusterctl/client/config/reader_memory.go

```diff
@@ -0,0 +1,146 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/yaml"
+)
+
+// MemoryReader provides a reader implementation backed by a map.
+// This is to be used by the operator to place config from a secret
+// and the ProviderSpec.Fetchconfig.
+type MemoryReader struct {
+	variables   map[string]string
+	providers   []configProvider
+	imageMetas  map[string]imageMeta
+	certManager serializableCertManager
+}
+
+var _ Reader = &MemoryReader{}
+
+type serializableCertManager struct {
+	URL     string `json:"url,omitempty"`
+	Version string `json:"version,omitempty"`
+	Timeout string `json:"timeout,omitempty"`
+}
+
+// NewMemoryReader return a new MemoryReader.
+func NewMemoryReader() *MemoryReader {
+	return &MemoryReader{
+		variables:  map[string]string{},
+		imageMetas: map[string]imageMeta{},
+		providers:  []configProvider{},
+	}
+}
+
+// Init initialize the reader.
+func (f *MemoryReader) Init(_ string) error {
+	data, err := yaml.Marshal(f.providers)
+	if err != nil {
+		return err
+	}
+	f.variables["providers"] = string(data)
+	data, err = yaml.Marshal(f.imageMetas)
+	if err != nil {
+		return err
+	}
+	f.variables["images"] = string(data)
+	return nil
+}
+
+// Get gets a value for the given key.
+func (f *MemoryReader) Get(key string) (string, error) {
+	if val, ok := f.variables[key]; ok {
+		return val, nil
+	}
+	return "", errors.Errorf("value for variable %q is not set", key)
+}
+
+// Set sets a value for the given key.
+func (f *MemoryReader) Set(key, value string) {
+	f.variables[key] = value
+}
+
+// WithVar sets the map entry key=value.
+func (f *MemoryReader) WithVar(key, value string) *MemoryReader {
+	f.variables[key] = value
+	return f
+}
+
+// UnmarshalKey gets a value for the given key, then unmarshal it.
+func (f *MemoryReader) UnmarshalKey(key string, rawval interface{}) error {
+	data, err := f.Get(key)
+	if err != nil {
+		return nil // nolint:nilerr // We expect to not error if the key is not present
+	}
+	return yaml.Unmarshal([]byte(data), rawval)
+}
+
+// WithProvider adds the given provider to the "providers" map entry and returns any errors.
+func (f *MemoryReader) WithProvider(name string, ttype clusterctlv1.ProviderType, url string) (*MemoryReader, error) {
+	f.providers = append(f.providers, configProvider{
+		Name: name,
+		URL:  url,
+		Type: ttype,
+	})
+
+	yaml, err := yaml.Marshal(f.providers)
+	if err != nil {
+		return f, err
+	}
+	f.variables["providers"] = string(yaml)
+
+	return f, nil
+}
+
+// WithProviderUnsafe adds the given provider to the "providers" map entry.
+// Note: any errors are ignored.
+func (f *MemoryReader) WithProviderUnsafe(name string, ttype clusterctlv1.ProviderType, url string) *MemoryReader {
```

Comment:
As far as I remember this originated by following considerations:
- MemoryReader is equal to FakeReader, let's dedup
- MemoryReader, being used on production, can't ignore errors like we were doing in the fake reader
- Let's make this evident by adding Unsafe variants to be used in tests.

However, reviewing this after some time it seems to me that we are promoting to a first level API some concern that should be limited to tests only. 

Wondering if we should rollback on this and keep the FakeReader around with methods ignoring errors (or panicking in case of errors), so we can have a cleaner API surface for the MemoryReader with the Add provider method only.
Ideally, FakeReader should wrap the memory reader, but I'm not sure if this will lead to circular dependencies

@asalkeld @sbueringer opinions?

Reply:
Sounds reasonable. Right now it's not really ideal.

Reply:
Sure, I'll update. There is a circular import cycle, so I can't use MemoryReader from FakeReader unfortunately.